NuFast oscillation calculator

cullenmsullivan opened the pull request:
This PR incorporates an implementation of the [NuFast](https://arxiv.org/abs/2405.02400) algorithm by Peter Denton (FNAL) and Stephen Parke (BNL). The algorithm is a 3F unitarity-assuming no-NSI neutrino oscillation calculator optimized for long baseline experiments. The algorithm is similar in spirit to `OscCalcAnalytic`, except some approximations make finding eigenvectors of the Hamiltonian slightly faster.
Testing the algorithm, so far I have found that although Stan fits are not faster with `OscCalcNuFast` compared to `OscCalcAnalytic`, I have found that calculations for the `<double>` template are 20%~30% faster (see NOvA docdb-64478 and docdb-64815).
In this pull request, I have migrated the caching system of `OscCalcAnalytic` to a new file, `Cache.h` so that it can be shared between `OscCalcAnalytic` and `OscCalcNuFast`.

chenel commented on OscLib/OscCalcNuFast.cxx:
```diff
@@ -0,0 +1,295 @@
+    
+    // NOvA doesn't use dmsq31 by convention, but we can compute it if we know dmsq32 and dmsq21.
+    const T
+    Dmsq31 = Dmsq32 + Dmsq21, // dmsq31 = msq3 - msq1 = (msq3-msq2) + (msq2-msq1) = dmsq32 + dmsq21.
```
Might be worth a comment here that this works for both mass orderings because in IO `Dmsq32` and `Dmsq21` have opposite sign

chenel commented on OscLib/OscCalcNuFast.h:
```diff
@@ -0,0 +1,94 @@
+    }
+    /// Copy constructor.
+    _OscCalcNuFast(const _OscCalcNuFast<T>& other);
+    /// from and to aren't necessary for computing the matrix but are used for the return (so the cache doesn't need to be searched for the result).
+    template<class VT, class KVT> inline __attribute__((always_inline)) VT RecomputeProbabilityMatrix(int from, int to, KVT E);
+    // Master probability function called with some combination of primitives, stan::math::vars, and Eigen objects.
+    template<class VT, class KVT> inline __attribute__((always_inline)) VT _P(int from, int to, KVT E);
```
Force-inlining these two quite-complex functions seems fairly dubious to me.  They are long enough---and `RecomputeProbabilityMatrix()` can even do an allocation in it (!)---that I would be surprised if inlining made any measurable difference.  Did you do any performance checking to verify that it helps?  Did you do any comparisons on the size of the output code to see if there was any impact?

cullenmsullivan replied:
That's a good point. I didn't think very hard about aggressively inlining these functions because `OscCalcAnalytic` does similar things.
I didn't make any testing with and without inlining, but I highly doubt it's actually helpful in any measurable way. I want to remove the inlining and assume that it makes no change to runtime.

chenel commented on OscLib/OscCalcAnalytic.h:
```diff
@@ -1,18 +1,19 @@
 #include "OscLib/IOscCalc.h"
 
 #include <functional>
-#include <unordered_map>
+//#include <unordered_map>
```
just remove the line, rather than commenting out.  we can always resurrect later from Git if needed

cullenmsullivan replied:
Oops, this is an oversight from me. I forgot to tidy `OscCalcAnalytic` before making the PR - will fix.

chenel commented on OscLib/OscCalcAnalytic.h:
```diff
@@ -1,18 +1,19 @@
 }
 
-// We want to put ArrayXd into an unordered_map, so define hash and equality
+/*// We want to put ArrayXd into an unordered_map, so define hash and equality
```
similarly (just delete, don't comment)

chenel commented on OscLib/OscCalcAnalytic.h:
```diff
@@ -81,7 +82,7 @@ namespace osc::analytic
   };
 
-  template<class T> class Probs
+  /*template<class T> class Probs
```
and again (just delete, don't comment)

chenel commented on OscLib/OscCalcNuFast.cxx:
```diff
@@ -0,0 +1,295 @@
+    {}
+  
+  template<typename T>
+  _OscCalcNuFast<T>::~_OscCalcNuFast(void) {}
```
in C++11 an `~_OscCalcNuFast() = default;` in the class declaration is preferred to an empty explicit destructor implementation

chenel commented on OscLib/OscCalcNuFast.cxx:
```diff
@@ -0,0 +1,295 @@
+  _OscCalcNuFast<T>::~_OscCalcNuFast(void) {}
+  
+  template<typename T>
+  _OscCalcNuFast<T>* _OscCalcNuFast<T>::Copy(void) const { return new _OscCalcNuFast<T>(*this); }
```
`(void)` is a C-ism that's generally frowned upon in C++ for ideological reasons.  not a big deal, but looks weird
```suggestion
  _OscCalcNuFast<T>* _OscCalcNuFast<T>::Copy() const { return new _OscCalcNuFast<T>(*this); }
```

cullenmsullivan replied:
Actually, I prefer using `(void)`, and I wasn't aware of the modern C++ convention avoiding it. It makes more sense to me to write it with `(void)`, but this isn't a hill I care do die on or anything.

chenel commented on OscLib/OscCalcNuFast.cxx:
```diff
@@ -0,0 +1,295 @@
+    // ---------------------------------------------------------------------------- //
+    // Newton iterations to improve lambda3 arbitrarily, if needed, (B needed here) //
+    // ---------------------------------------------------------------------------- //
+    const VT B = Dmsq21*Dmsq31 + Amatter * See; // B is only needed for N_Newton >= 1
+    for (int i = 0; i < N_Newton; i++) {
+      lambda3 = (lambda3 * lambda3 * (lambda3 + lambda3 - A) + C) / (lambda3 * (2 * (lambda3 - A) + lambda3) + B); // this strange form prefers additions to multiplications
+    }
+    
```
If this block is only used sometimes, it should be inside an `if` so that the derivatives aren't computed by Stan when not used

cullenmsullivan replied:
I stuck this block inside an `if` and confirmed that the spectra still converge. Actually because I got curious, I just discovered that there is no (well, some extremely small) difference between 0 and 1 Newton iterations by testing spectra. I'm going to set the default to 0 Newton iterations.

chenel commented on OscLib/OscCalcNuFast.h:
```diff
@@ -0,0 +1,94 @@
+    /// Electron density in matter. NuFast recommends 0.5, and this choice matches other calcs.
+    double fYe;
+    /// Number of Newton steps to use to improve eigen{value}{vector} slns. Increase me for more accuracy and worse runtime.
+    double fNNewton;
```
why is this guy a `double` and not an `unsigned int`, out of curiosity?

cullenmsullivan replied:
This is simply a mistake from me - good catch!

chenel reviewed (commented):
In general, looks very nice, and is fine for a first implementation.  Most of my inline comments are simple ones.

Bigger question: I've only looked over `OscCalcNuFast` so far.  Is `OscCalcNuFastDumb` something you actually want to include in the final product?  It seems to be the same apart from the caching?

I have a couple of meta-thoughts that we should discuss, as well:

1. Rather than copy-pasting-editing the code from NuFast, it would be much neater if there were a `nufast::Calculator` object we could just make a member of the `OscLib` object and use it, thus avoiding duplicating the code from them.  Obviously the NuFast authors didn't structure their code this way, but I wonder if maybe they would be interested in doing such a thing if they knew we were using it.  And if so, I bet your templated version would not be a difficult sell.  I know Peter Denton and would be happy to initiate the conversation.
2. I'm a bit uneasy about the cache when applied to Stan types.  Stan uses an internal buffer for storing its autodiff stack that's not exposed to the user, and it can be cleared without warning (if the user does anything that involves actually computing the derivatives).  I've added guardrails to the usage with `StanFitter`, so that this is less likely to happen in practice, but we should maybe talk about how necessary the caching actually is.  (And all of this obviously applies to `OscCalcAnalytic`, which nobody has seriously used so far.)

cullenmsullivan commented:
I've addressed most of the inline changes. To address the larger issues:

1. Including `OscCalcNuFastDumb` in this PR is a mistake from me, and I never intended to actually commit it. It's strictly slower than `OscCalcNuFast` as-is, and I only used it to help benchmark my progress as I increasingly increased how aggressively things are cached. I've now removed it. With `OscCalcNuFastDumb` gone, there's no immediate need to refactor `OscCalcNuFast`, but it could improve the readability and extensibility (perhaps NSI/sterile folks would be interesting in adapting it?). When writing this, I tried to balance compatibility with `OscLib` but also preservation of the original algorithm as the authors wrote it. I'm open to collaboration with the authors to streamline the code more.
2. Avoiding use of the cache with Stan types would be an easy change, but from a very quick test, it seems that the speed of Stan MCMC sampling is sensitive to use of the cache. When running a fit side-by-side with and without a caching calculator, the caching fit ran about twice as quickly. This might need more study. Because `StanFitter` uses the `Eigen` templates for getting oscillation probabilities, the calculator shouldn't need to be queried for energy multiple times for one channel (right?). Perhaps the extra time is going into recomputing probabilities for multiple flavor -> flavor channels? If this is the case, we might be able to avoid the slowdown by saving all disappearance/appearance channels at the fitter level and also avoid use of a cache for Stan vars at the calculator level?

chenel commented:
Thanks for studying the cache usefulness.  I think that given that it seems to matter, let's keep it for now and try to add guardrails later if we find people using the Stan-aware calculator irresponsibly. 😄   If you could add a warning in the comments, that would be great.

Let's write to Peter and get a feel for his level of interest in extending the code he has.  (The main advantage I can see of modularizing is if in the future he and his collaborators make some improvement, we may be able to adopt it without any tinkering on our end.)  I'm happy to start that conversation.

chenel commented:
In the meantime, I think this is ready for regular review and incorporation into `OscLib`, so you can feel free to promote this out of 'draft' status.

cullenmsullivan commented:
All right, I've marked it as ready for review. I think talking to Peter is a good idea as well - I'd appreciate it if you do start the conversation, thank you!

Following up from this PR (once merged), I'll make a sister PR in novasoft to save my testing macros and make an edit to `StanFitter` to incorporate the Stan-aware calculator.

chenel reviewed (dismissed):
I'm happy with merging, per discussion above.

vhewes commented on OscLib/OscCalcAnalytic.cxx:
```diff
@@ -344,7 +344,7 @@ namespace osc::analytic
     }
   }
-
+*/
```
since this function is now pulled in from `Cache.h`, this function should be deleted instead of just commented out

cullenmsullivan replied:
Oops, that's another piece I forgot to tidy - thanks for the good catch. I just committed a deletion

vhewes reviewed (changes requested):
overall this looks good, i just have one change to request. i definitely agree with @chenel's comment about pulling this code in from NuFast if that's possible

gavinsdavies commented:
this just needs @vhewes to give approval to the requested changes.

vhewes reviewed (approved):
looks good! sorry for the delay
